cmd/trcd/main.go

```diff
@@ -29,31 +39,96 @@ func init() {
 func main() {
 	flag.Parse()
 
-	conn, err := net.Listen("unix", *sock)
+	if *debug {
+		log.SetLevel(log.DebugLevel)
+	}
+
+	var netLst net.Listener
+	var err error
+	if *useTCP {
+		log.Info("Listening to TCP socket...")
+		addr, _ := net.ResolveTCPAddr("", ":"+strconv.Itoa(internalPort))
+		netLst, err = net.ListenTCP("tcp", addr)
+
+	} else {
+		log.Info("Listening to Unix socket...")
+		netLst, err = net.Listen("unix", *sock)
+	}
 	if err != nil {
 		panic(err)
 	}
-	defer conn.Close()
+	defer netLst.Close()
 
 	for {
-		c, err := conn.Accept()
+		sockConn, err := netLst.Accept()
+		log.Infof("Connected to %s", sockConn.RemoteAddr())
 		if err != nil {
 			panic(err)
 		}
-		defer c.Close()
+		defer sockConn.Close()
+
+		// state handler of mock TRC, logs all actions
+		setStateHandler := func(msg *api.Message) (*api.Message, error) {
+			log.Infof("Received: %s", msg)
+
+			reply, err := trctest.DefaultStateHandler(msg)
+			log.Debugf("Reply with: %s", reply)
+			return reply, err
+		}
+		// ping handler of mock TRC, logs when ping is received
+		pingHandler := func(msg *api.Message) (*api.Message, error) {
+			log.Debug("Received ping")
+			return trctest.DefaultPingHandler(msg)
+		}
+
+		trcConn := trctest.Connect(sockConn, sockConn,
+			trctest.WithHandler(api.MessageTypeState, setStateHandler),
+			trctest.WithHandler(api.MessageTypePing, pingHandler),
+			trctest.WithHandler(api.MessageTypeHandshake, trctest.DefaultHandshakeHandler),
+		)
 
 		go func() {
-			dec := json.NewDecoder(c)
-			dec.DisallowUnknownFields()
-			var msg api.Message
-			if err := dec.Decode(&msg); err != nil {
-				log.Printf("Failed to decode input: %s", err)
+			for err := range trcConn.Errors() {
+				log.WithError(err).Error("Internal mock-trc error")
 			}
-			log.Printf("Got: %v", msg)
 		}()
-		if _, err := io.Copy(c, os.Stdin); err != nil {
-			log.Printf("Failed to copy input: %s", err)
+
+		if err := trcConn.SendHandshake(&api.Handshake{Version: trcapi.DefaultVersion}); err != nil {
+			panic(err)
+		}
+		log.Debug("Received handshake")
+
+		if err := trcConn.SendState(&api.State{}); err != nil {
+			log.Error("initial state: " + err.Error())
 			return
 		}
+
+		closeCh := make(chan struct{})
+		defer close(closeCh)
+
+		// send random messages unless silent is set
+		if !(*silent) {
+			go func() {
+				hBeat := time.Tick(3 * time.Second)
+				for {
+					aWhile := time.Millisecond * time.Duration(rand.Intn(5000))
+					// wait 0-5 seconds, then send an empty state
+					select {
+					case <-time.After(aWhile):
+						if err := trcConn.SendState(apitest.RandomState()); err != nil {
+							log.Error("Random sendState error: " + err.Error())
+						}
```

Comment:
There should be an `else` here, otherwise you print both the error and the `Sent`